### ISSUE TITLE
TSDB (histograms): Counter reset header might go undetected during ingest

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#15358]
tl;dr: If we cut a new chunk because the schema or the zero threshold has changed in a "compatible" way, we will always set a counter reset header of "no counter reset" in the new chunk although we haven't really checked if there was a counter reset. Luckily, this isn't hurting us right now because we don't trust "no counter reset" in the counter reset header anyway (for other reasons).

More detailed explanation:

First of all: What is a "compatible" change of the schema or the zero threshold? This happens when the instrumented target decides to get rid of some buckets by reducing the resolution and/or grow the zero bucket. The previous histograms (with the higher resolution and/or the smaller zero bucket) can be converted to the new schema and zero bucket width so the histograms are still compatible with each other for PromQL operations.

The "gold standard" for counter reset detection is the [FloatHistogram.DetectReset](https://github.com/prometheus/prometheus/blob/fc3d19725d83e13363ce40d56672c6f97e2955e7/model/histogram/float_histogram.go#L604) method. If it encounters a difference in schema and/or the zero threshold, it will find out if the change is compatible (see above) and will perform the conversion for the earlier histogram first, after which it will run the full reset detection (checking all bucket counts and the total count).

During ingestion, though, we are mostly interested if we have to cut a new chunk at all. One of the reason is a counter reset, but _any_ change in schema requires cutting a new chunk. So if there is a change in schema, the code quickly decides that a new chunk has to be cut, without further counter reset detection. This is fine, in principle, but the problem is that we set a "no counter reset" counter reset header of the new chunk in this case, which might or might not represent the truth. Instead, we either need to run a full counter reset detection or simply set a counter reset header of "unknown counter reset".

Given that we currently do not trust a "no counter reset" header anyway (because the preceding chunk could get removed, which would invalidate the information anyway), this is not breaking anything right now, but we should still fix the behavior. I propose for now to just set "unknown counter reset" in this case because investing all the work in a full counter reset detection is in vain as long as we have no way of trusting a "no counter reset" header (which is something we might do in the future, see https://github.com/prometheus/prometheus/issues/15346 for ideas). Most likely, we should change the [appendable method](https://github.com/prometheus/prometheus/blob/fc3d19725d83e13363ce40d56672c6f97e2955e7/tsdb/chunkenc/histogram.go#L265) to return a `CounterResetHeader` rather than just a bool.

/cc @fionaliao @krajorama

**Comment by fatsheep9146** [comment]
Could I take this issue?

**Comment by beorn7** [comment]
Sure. Thanks @fatsheep9146 . I have assigned you to this issue.
